app/Jobs/CreateQueryserviceBatchesJob.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+namespace App\Jobs;
+
+use App\EventPageUpdate;
+use App\QsBatch;
+use App\QsCheckpoint;
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Config;
+
+class CreateQueryserviceBatchesJob extends Job
+{
+    private const NAMESPACE_ITEM = 120;
+    private const NAMESPACE_PROPERTY = 122;
+    private const NAMESPACE_LEXEME = 146;
+
+    private int $entityLimit;
+
+    public function __construct()
+    {
+        $this->entityLimit = Config::get('wbstack.qs_batch_entity_limit');
+    }
+
+    public function handle(): void
+    {
+        DB::transaction(function () {
+            $lastCheckpoint = QsCheckpoint::get();
+
+            $newEntities = $this->getNewEntities($lastCheckpoint);
+            foreach ($newEntities as $wikiId => $entityIdsFromEvents) {
+                $ok = $this->tryToAppendEntitesToExistingBatches($entityIdsFromEvents, $wikiId);
+                if ($ok) {
+                    continue;
+                }
+                $this->createNewBatches($entityIdsFromEvents, $wikiId);
+            }
+
+            QsCheckpoint::set($this->getLatestEventId($lastCheckpoint));
+        });
+    }
+
+    private function getLatestEventId(int $lastCheckpoint): int
+    {
+        $next = EventPageUpdate::where(
```

Comment:
This also means we trade an additional DB query against removing some slightly brittle looping logic.